### PR TITLE
Add driver for pgx v5

### LINKED ISSUE #827: Question: migrate or implement pgx v5 driver
**Is your feature request related to a problem? Please describe.**

PGX driver V5 was released - https://pkg.go.dev/github.com/jackc/pgx/v5. 

**Describe the solution you'd like**

1. I suggest migrating from pgx v4 to pgx v5 OR write new `PgxV5` driver implementation
2. Update pgx4 module to latest 4.17.2, if need to have v4 driver

### PR DESCRIPTION
**Pull request by treuherz**
Fixes #827

- Move existing driver into database/pgx/v4
- Make exported items into database/pgx into links to database/pgx/v4,
  marked as Deprecated
- Mark defunct ErrDatabaseDirty as Deprecated - no code returns it
- Bump pgx/v4 dependency
- Add test runs for newer Postgres versions

If you'd like the last couple of changes to be broken into a different PR then let me know.

### CODE REVIEW

**Comment by koenbollen** [comment]
Hey! Quick bump from me? I'm currently running this branch to allow me to use pgx and golang-migrate together. (thanks btw!)

Any updates on getting this merged? 😇

**Comment by treuherz** [comment]
Rebased on top of master

**Comment by mprimeaux** [comment]
Any updates on getting this merged? Appreciate everyone's time.

**Comment by pmorelli92** [comment]
Please some prio 🙏

**Comment by franchb** [comment]
Any updates on this PR please?

**Review by vincentdaniel — APPROVED** [review]

**Comment by MNovruzov** [comment]
Any updates please???

**Comment by mprimeaux** [comment]
Hi @dhui. Would you mind reviewing and consider an approval?  We have quite a few platforms waiting for this PR to be merged. Your help, and the help of any maintainer,  is greatly appreciated.

**Comment by treuherz** [comment]
Rebased on top of master again to fix conflicts, and bumped pgx minor versions again.

@vincentdaniel  it was nice of you to approve this last time, but it didn't look like that granted mergeability, is that correct?

File: database/pgx/pgx.go
Comment (dhui) [inline database/pgx/pgx.go:36]: These would need to be exported as well.
I think it's safer and easier to leave the default as v4 (e.g. remove the v4 folder in this PR) and update the README and godocs to also mention that a v5 driver is also available. I think it's fine to offer both v4 and v5 so we can mark v4 as deprecated later. e.g. whenever pgx deprecates v4. Do you know if v4 has already been deprecated?

File: database/pgx/v5/README.md
Comment (dhui) [inline database/pgx/v5/README.md:3]: This will need to be updated

File: internal/cli/build_pgx.go
Comment (dhui) [inline internal/cli/build_pgx.go:7]: 🎉

**Review by dhui — CHANGES_REQUESTED** [review]
@treuherz thanks for the PR and apologies for the slow turnaround time. Life has been busy...

I'll also update golang-migrate to only support (and run tests for) Go 1.19 and 1.20

**Comment by treuherz** [comment]
Responded to comments, rebased on master again to resolve go.mod conflicts.

File: README.md
Comment (dhui) [inline README.md:27]: This should be `database/pgx`

File: database/pgx/v5/pgx.go
Comment (dhui) [inline database/pgx/v5/pgx.go:30]: This should be registered as `pgx5` to avoid a conflict with `pgx` (v4)

File: internal/cli/build_pgxv5.go
Comment (dhui) [inline internal/cli/build_pgxv5.go:1]: Can you rename this file and tag to `pgx5`? The `v` is verbose and unnecessary.

File: Makefile
Comment (dhui) [inline Makefile:2]: You'll also need to update this to `pgx5`

**Review by dhui — CHANGES_REQUESTED** [review]
@treuherz Thanks for the PR updates. This _should_ be the last round.

File: database/pgx/v5/pgx.go
Comment (treuherz) [reply database/pgx/v5/pgx.go:30]: As of a few weeks ago, pgx registers it’s database/sql Driver under `pgx/v4` or `pgx/v5`, as well as conditionally registering `pgx` if it’s not already registered by importing another version (see https://github.com/jackc/pgx/blob/9ae852eb583d2dced83b1d2ffe1c8803dda2c92e/stdlib/sql.go#L94). I’m happy to copy in the conflict-avoidance code or if you prefer just leave it out and have v4 register as pgx and pgx/v4, and v5 only register pgx/v5

I definitely think it would be better to match pgx’s `pgx/vN` registration string rather than inventing a new one. I know I didn’t use it in this PR originally but the upstream change is very new!

File: database/pgx/pgx.go
Comment (dhui) [inline database/pgx/pgx.go:31]: Register this as `pgx4` since `pgx/v4` is verbose. We don't need to be consistent with the upstream. Also, our DSN/connection string parsing uses URL parsing and `/` is not a [valid scheme character](https://datatracker.ietf.org/doc/html/rfc3986#section-3.1)

File: database/pgx/v5/pgx.go
Comment (dhui) [inline database/pgx/v5/pgx.go:30]: Register as `pgx5` for the same reason

**Review by dhui — CHANGES_REQUESTED** [review]
@treuherz Could you also update the db registration name?

**Comment by kriswill** [comment]
can we merge/release this?

**Review by dhui — APPROVED** [review]
@treuherz Thanks again for the PR and for addressing my feedback!

**Pull request merged**
